# Break Points - Version Pre-Alpha

## Program Functionality

The main purpose of this pre-alpha version of the program is to identify specific targets of a given endonuclease based on RNA-seq data. For accurate results, your experiment should meet the following criteria:

1. The target molecule must be an endonuclease.
2. The data must come from a knockout experiment.
3. While not mandatory, it's advisable that your sequencing data be the result of a library that doesn't fragment RNA.

The Break Points algorithm scans through all positions of your sequencing data, compares read counts, identifies cleavage points (break points), and then compares the control strain with the mutant (lacking the target endonuclease) to determine if the break points are influenced by the knocked-out endonuclease.

A **break point** is defined as a position in the genome where there is a high read count (minimum 10) in the 5' profile data, followed by, within a maximum interval of 3 base pairs, a high read count position in the 3' profile, indicating a cleavage point. The algorithm also performs this analysis backward, from the 3' profile to the 5' profile, but this information is irrelevant for single-end data.

This program complements the [frtc pipeline](https://github.com/alanlorenzetti/frtc), which processes raw RNA-seq data. Break Points is designed to receive bedgraph profile files as input, indicating counts in starting and ending read positions for a given transcript segment.

If the break point position has a log2 fold change = 1 difference, with the control strain having more read counts than the mutant, the program adds the position as influenced by the target endonuclease.

## Usage

This program requires the R environment to be installed for proper functioning.

To use the program, open the Linux command line, and input the following command with 8 arguments, one of which is optional:

```bash
R brk.R <Control Strain 5' profile> <Mutant Strain 5' profile> <Control Strain 3' profile> <Mutant Strain 3' profile> <Control Coverage> <Mutant Coverage> <Chromosome> <Output Name>
```

The first 6 arguments refer to the .bedgraph files generated by the frtc pipeline. The Chromosome argument specifies the chosen chromosome, and the Output Name argument is optional, allowing you to change the name of the output files. If not set, the files will be named "out".

It's recommended to save the variables in the command line before using brk.R for cleanliness.

Break Points will create two folders: "Raw_Tables" (containing main tables with data information) and "Bedgraph_Version" (containing bedgraph tables ready to use in Integrative Genome Browser - IGV).

## Further Information
Contact me via email, jpfacio@usp.br, if you have any questions or have used Break Points in your work.
